src/platform/intel/cavs/lib/clk.c

```diff
@@ -101,24 +100,69 @@ void platform_clock_on_wakeup(void)
 		set_cpu_current_freq_idx(freq_idx);
 	}
 }
+#endif
 
-void platform_clock_on_waiti(void)
+void platform_clock_waiti_entry(void)
 {
-	int freq_idx = get_cpu_current_freq_idx();
+	int freq_idx;
+	int target_idx;
+
+	/* update the active clock according to the pm runtime state */
+	if (pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID) &&
+	    *cache_to_uncache(&active_freq_idx) != CPU_HPRO_FREQ_IDX)
+		platform_set_active_clock(CPU_HPRO_FREQ_IDX);
+	else if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX)
```

Comment:
Shouldn't you check for !pm_runtime_is_active here? If  pm_runtime_is_active returns true and active clock is already set to HPRO, I believe we'd land in that else if block and put the DSP back in LPRO wrongly